resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,575 @@
+	 * @return a supplier which is decorated by a Bulkhead.
+	 */
+	static <T> Callable<T> decorateCallable(AdaptiveBulkhead bulkhead, Callable<T> callable) {
+		return () -> {
+			long start = 0;
+			boolean isFailed=false;
+			bulkhead.acquirePermission();
+			try {
+				start = Instant.now().toEpochMilli();
+				return callable.call();
+			} catch (Exception e) {
+				isFailed=true;
+				bulkhead.onError( start,TimeUnit.MILLISECONDS, e);
+				throw e;
+			} finally {
+				if (start != 0 && !isFailed) {
+					bulkhead.onSuccess(start, TimeUnit.MILLISECONDS);
+				}
+			}
+		};
+	}
+
+	/**
+	 * Returns a supplier which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param supplier the original supplier
+	 * @param <T>      the type of results supplied by this supplier
+	 * @return a supplier which is decorated by a Bulkhead.
+	 */
+	static <T> Supplier<T> decorateSupplier(AdaptiveBulkhead bulkhead, Supplier<T> supplier) {
+		return () -> {
+			long start = 0;
+			boolean isFailed=false;
+			bulkhead.acquirePermission();
+			try {
+				start = Instant.now().toEpochMilli();
+				return supplier.get();
+			} catch (Exception e) {
+				isFailed=true;
+				bulkhead.onError( start,TimeUnit.MILLISECONDS, e);
+				throw e;
+			} finally {
+				if (start != 0 && !isFailed) {
+					bulkhead.onSuccess(start, TimeUnit.MILLISECONDS);
+				}
+			}
+		};
+	}
+
+	/**
+	 * Returns a consumer which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param consumer the original consumer
+	 * @param <T>      the type of the input to the consumer
+	 * @return a consumer which is decorated by a Bulkhead.
+	 */
+	static <T> Consumer<T> decorateConsumer(AdaptiveBulkhead bulkhead, Consumer<T> consumer) {
+		return t -> {
+			long start = 0;
+			boolean isFailed=false;
+			bulkhead.acquirePermission();
+			try {
+				start = Instant.now().toEpochMilli();
+				consumer.accept(t);
+			} catch (Exception e) {
+				isFailed=true;
+				bulkhead.onError( start,TimeUnit.MILLISECONDS, e);
+				throw e;
+			} finally {
+				if (start != 0 && !isFailed) {
+					bulkhead.onSuccess(start, TimeUnit.MILLISECONDS);
+				}
+			}
+		};
+	}
+
+	/**
+	 * Returns a consumer which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param consumer the original consumer
+	 * @param <T>      the type of the input to the consumer
+	 * @return a consumer which is decorated by a Bulkhead.
+	 */
+	static <T> CheckedConsumer<T> decorateCheckedConsumer(AdaptiveBulkhead bulkhead, CheckedConsumer<T> consumer) {
+		return t -> {
+			long start = 0;
+			boolean isFailed=false;
+			bulkhead.acquirePermission();
+			try {
+				start = Instant.now().toEpochMilli();
+				consumer.accept(t);
+			} catch (Exception e) {
+				isFailed=true;
+				bulkhead.onError( start,TimeUnit.MILLISECONDS, e);
+				throw e;
+			} finally {
+				if (start != 0 && !isFailed) {
+					bulkhead.onSuccess(start, TimeUnit.MILLISECONDS);
```

Comment:
Why do we need `onSuccess` in the finally part?
Shouldn't it be the total duration instead of `start`?
`long durationInNanos = System.nanoTime() - start;`

Reply:
yes it should be duration , updated accordingly, we need on success on finally block as we need record the successful calls if they are not failed one !